### PR TITLE
[PR 2] Issue-523: Replace go-micro with gRPC framework

### PR DESCRIPTION
**Pull request by a-ajith**
The PR is for issue 523 task 2: Add etcd to the Kubernetes deployment

The PR contains the changes required for building etcd image

### CODE REVIEW

**Review by Bharath-KKB — APPROVED** [review]

File: install/Docker/dockerfiles/Dockerfile.etcd [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+WORKDIR /etcd/
+
+# userid and groupid of odimra user and odimra group respectively
+# which will present be on the host machine.
+ARG ODIMRA_USER_ID
+ARG ODIMRA_GROUP_ID
+
+RUN apt-get update -q=3 && apt-get -q install -q=3 -y \
+        vim \
```
Comment (bharath-b-hpe) [inline install/Docker/dockerfiles/Dockerfile.etcd]: Please remove vim package it's not required and it increases the image size.

File: install/Docker/dockerfiles/Dockerfile.etcd [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        wget \
+        && apt-get clean \
+        && rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+RUN if [ -z "$ODIMRA_USER_ID" ] || [ -z "$ODIMRA_GROUP_ID" ]; then echo "\n[$(date)] -- ERROR -- ODIMRA_USER_ID or ODIMRA_GROUP_ID is not set\n"; exit 1; fi \
+	&& mkdir -p /opt/etcd/bin /opt/etcd/data /opt/etcd/conf /opt/etcd/scripts /etcd/tmp \
+        && export ETCD_IMAGE_URL=https://storage.googleapis.com/etcd/v3.4.0/etcd-v3.4.0-linux-amd64.tar.gz \
+        && export ETCD_IMAGE_BUNDLE=etcd-v3.4.0-linux-amd64.tar.gz \
```
Comment (bharath-b-hpe) [inline install/Docker/dockerfiles/Dockerfile.etcd]: Please check if 3.4.0 is the latest stable version.

File: install/Docker/dockerfiles/Dockerfile.etcd [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        && rm -rf /etcd \
+        && groupadd -r -g $ODIMRA_GROUP_ID odimra \
+	&& useradd -s /bin/bash -u $ODIMRA_USER_ID -m -d /home/odimra -r -g odimra odimra \
+	&& chown -R odimra:odimra /opt/etcd \
+	&& chmod 0700 /opt/etcd/data
+
+#COPY --chown=etcd:etcd ./scripts/* /opt/etcd/scripts/
+# RUN mv /opt/etcd/scripts/etcd.conf.tmpl /opt/etcd/bin/etcd.conf \
+# 	&& chmod 0700 /opt/etcd/scripts/* \
+# 	&& chmod 0600 /opt/etcd/bin/etcd.conf
```
Comment (bharath-b-hpe) [inline install/Docker/dockerfiles/Dockerfile.etcd]: Remove commented code

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
Address BharathBs' comments

File: odim-controller/helmcharts/etcd-ha/values.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+odimra:
+  namespace:
+  groupID:
+  etcdDirPath:
+  etcdImageTag: "1.16"
+  etcdNodePort:
```
Comment (bharath-b-hpe) [inline odim-controller/helmcharts/etcd-ha/values.yaml]: etcdNodePort is required? and also remove unused params if any.

File: odim-controller/helmcharts/etcd-ha/templates/etcd-ha-statefulset.yaml [hunk 5]
```diff
@@ -0,0 +1,83 @@
+apiVersion: apps/v1
```
Comment (bharath-b-hpe) [inline odim-controller/helmcharts/etcd-ha/templates/etcd-ha-statefulset.yaml:2]: etcd-ha-statefulset.yaml contains manifest for just one instance, please check.

File: odim-controller/odimra/group_vars/all [hunk 6]
```diff
@@ -23,11 +23,13 @@ odim_third_party_helmcharts:
    redis: redis-5.0.8-0.1.0
    kafka: kafka-2.5.0-0.1.0
    consul: consul-1.6.0-0.1.0
+   etcd: etcd-1.16.0-0.1.0
 odim_third_party_ha_helmcharts:
    zookeeper: zookeeper-ha-2.5.0-0.1.0
    redis: redis-ha-5.0.8-0.1.0
    kafka: kafka-ha-2.5.0-0.1.0
    consul: consul-ha-1.6.0-0.1.0
+   etcd: etcd-1.16.0-0.1.0
```
Comment (bharath-b-hpe) [inline odim-controller/odimra/group_vars/all]: this should be etcd-ha-1.16.0-0.1.0 right?

Comment (Bharath-KKB) [reply odim-controller/odimra/group_vars/all]: do we still need consul ? BB i think we have to stop the cluster and install. And this is one way upgrade unless part of initial install.

Comment (a-ajith) [reply odim-controller/odimra/group_vars/all]: @Bharath-KKB we will remove consul once all our services are moved from GoMicro to gRPC.

Comment (a-ajith) [reply odim-controller/odimra/group_vars/all]: @bharath-b-hpe Addressed the comment

File: odim-controller/helmcharts/etcd-ha/values.yaml (hunk 4 above)
Comment (a-ajith) [reply odim-controller/helmcharts/etcd-ha/values.yaml]: @bharath-b-hpe Addressed

File: install/Docker/dockerfiles/Dockerfile.etcd (hunk 3 above)
Comment (a-ajith) [reply install/Docker/dockerfiles/Dockerfile.etcd]: @bharath-b-hpe Addressed

File: install/Docker/dockerfiles/Dockerfile.etcd (hunk 1 above)
Comment (a-ajith) [reply install/Docker/dockerfiles/Dockerfile.etcd]: @bharath-b-hpe Addressed

File: odim-controller/helmcharts/etcd-ha/templates/etcd-ha-statefulset.yaml (hunk 5 above)
Comment (a-ajith) [reply odim-controller/helmcharts/etcd-ha/templates/etcd-ha-statefulset.yaml:2]: Addressed

File: install/Docker/dockerfiles/Dockerfile.etcd (hunk 2 above)
Comment (a-ajith) [reply install/Docker/dockerfiles/Dockerfile.etcd]: changed version to 3.4.15

**Review by Bharath-KKB — APPROVED** [review]

**Pull request merged**